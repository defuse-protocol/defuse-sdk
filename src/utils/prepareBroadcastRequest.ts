import { base58, base64 } from "@scure/base"
import type { WalletMessage, WalletSignatureResult } from "../types"

export function prepareSwapSignedData(
  signature: WalletSignatureResult,
  walletMessage: WalletMessage
) /* todo: type should be inferred from API schema */ {
  switch (signature.type) {
    case "NEP141": {
      return {
        standard: "nep141",
        message: walletMessage.NEP141.message,
        nonce: base64.encode(walletMessage.NEP141.nonce),
        recipient: walletMessage.NEP141.recipient,
        public_key: signature.signatureData.publicKey, // publicKey is already in the correct format
        signature: transformNEP141Signature(signature.signatureData.signature),
      }
    }
    case "EIP712": {
      throw new Error("EIP712 signature is not supported")
    }
    default:
      throw new Error("exhaustive check failed")
  }
}

function transformNEP141Signature(signature: string) {
  const encoded = base58.encode(base64.decode(signature))
  return `ed25519:${encoded}`
}
