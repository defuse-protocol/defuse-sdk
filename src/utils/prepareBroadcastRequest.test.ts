import { describe, expect, it } from "vitest"
import type { NEP141SignatureData, WalletMessage } from "../types"
import { prepareSwapSignedData } from "./prepareBroadcastRequest"

describe("prepareSwapSignedData()", () => {
  it("should return the correct signed data for a NEP141 signature", () => {
    const signature: NEP141SignatureData = {
      type: "NEP141",
      signatureData: {
        accountId: "user.near",
        publicKey: "ed25519:Gxa24TGbJu4mqdhW3GbvLXmf4bSEyxVicrtpChDWbgga",
        signature: "stH6JnShG+GwWCsfN9iu/m4un6qwYLN9Df+5oQYsL7Q=",
      },
    }

    const walletMessage: WalletMessage = {
      NEP141: {
        message: `{"foo":"bar"}`,
        recipient: "defuse.near",
        nonce: Buffer.from(
          "esXbbxyJNApGznX1v8kT5ojuat7jqUv84Ib+Q6hWdzI=",
          "base64"
        ),
      },
      EIP712: {
        json: "{}",
      },
    }

    expect(prepareSwapSignedData(signature, walletMessage)).toMatchSnapshot()
  })
})
