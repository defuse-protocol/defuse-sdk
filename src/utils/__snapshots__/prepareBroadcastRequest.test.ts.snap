// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`prepareSwapSignedData() > should return the correct signed data for a NEP141 signature 1`] = `
{
  "message": "{"foo":"bar"}",
  "nonce": "esXbbxyJNApGznX1v8kT5ojuat7jqUv84Ib+Q6hWdzI=",
  "public_key": "ed25519:Gxa24TGbJu4mqdhW3GbvLXmf4bSEyxVicrtpChDWbgga",
  "recipient": "defuse.near",
  "signature": "ed25519:D33GxHszz1ZpkXpnEWYAwkQxeYBU9YCkTMKZ5uwDS1B9",
  "standard": "nep141",
}
`;
